dipy/tracking/utils.py

```diff
@@ -921,6 +923,79 @@ def unique_rows(in_array, dtype='f4'):
     return in_array[diff_in_array]
 
 
+def calculate_cci(streamlines, max_mdf=5, subsample=12, power=1,
+                  override=False):
+
+    """ Computes the cluster confidence index (cci), which is an
+    estimation of the support a set of streamlines gives to
+    a particular pathway.
+
+    Ex: A single streamline with no others in the dataset
+    following a similar pathway has a low cci. A streamline
+    in a bundle of 100 streamlines that follow similar
+    pathways has a high cci.
+
+    See: Jordan et al. 2017
+    (Based on streamline MDF distance from Garyfallidis et al. 2012)
+
+    Parameters
+    ----------
+    streamlines : list of 2D (N, 3) arrays
+        A sequence of streamlines of length N (# streamlines)
+    max_mdf : int
+        The maximum MDF distance (mm) that will be considered a
+        "supporting" streamline and included in cci calculation
+    subsample: int
+        The number of points that are considered for each streamline
+        in the calculation. To save on calculation time, each
+        streamline is subsampled to subsampleN points.
+    power: int
+        The power to which the MDF distance for each streamline
+        will be raised to determine how much it contributes to
+        the cci. High values of power make the contribution value
+        degrade much faster. Example: a streamline with 5mm MDF
+        similarity contributes 1/5 to the cci if power is 1, but
+        only contributes 1/5^2 = 1/25 if power is 2.
+
+    Returns
+    -------
+    Returns an array of CCI scores
+
+    References
+    ----------
+    [Jordan17] Jordan K. Et al., Cluster Confidence Index: A Streamline‐Wise
+    Pathway Reproducibility Metric for Diffusion‐Weighted MRI Tractography,
+    Journal of Neuroimaging, vol 28, no 1, 2017.
+
+    [Garyfallidis12] Garyfallidis E. et al., QuickBundles a method for
+    tractography simplification, Frontiers in Neuroscience,
+    vol 6, no 175, 2012.
+
+    """
+
+    # error if any streamlines are shorter than 20mm
+    lengths = list(length(streamlines))
+    if np.array(lengths).min() < 20 and not override:
+        ValueError('Short streamlines found. We recommend removing them.'
+                   'To continue with short streamlines set override=True')
+
+    # calculate the pairwise MDF distance between all streamlines in dataset
+    subsamp_sls = set_number_of_points(streamlines, subsample)
+
+    cci_score_mtrx = np.zeros([len(subsamp_sls)])
+
+    for i, sl in enumerate(subsamp_sls):
+        mdf_mx = bundles_distances_mdf([subsamp_sls[i]], subsamp_sls)
+        if (1 * mdf_mx == 0).sum() > 1:
+            raise ValueError('Identical streamlines. CCI calculation invalid')
+        mdf_mx_oi = (mdf_mx > 0) & (mdf_mx < max_mdf) & ~ np.isnan(mdf_mx)
+        mdf_mx_oi_only = mdf_mx[mdf_mx_oi]
+        cci_score = np.sum(np.divide(1, np.power(mdf_mx_oi_only, power)))
```

Comment:
Any chance to have `mdf_mx_oi_only` equal to 0?  if yes, make sure to avoid the `DivisionZeroError`.

Reply:
I don't think so... here's my reasoning: 
- the mdf_mx can only be zero when two streamlines are identical (it's the mdf distance, so based on summing euclidean distances between subsampled points on streamlines.. this can't be negative so the only way to get to zero would be if all of the points were identical). 
- Since we raise a ValueError when there are identical streamlines, we should never get to the point where mdf_mx_oi_only since that's a subset of mdf_mx
- I added the ValueError for identical streamlines because, if you think about what the purpose of this outlier calculation is, it's not really telling you anything when streamlines are identical because they would give strong support to a pathway that was probably gained "unfairly"... e.g. 
1) this is being used for deterministic tracking with two seeds that are on top of each other, which wouldn't actually give you confidence in that pathway... only a seed that's a bit further away would be informative in a deterministic system... 2) it's really intended for probabilistic tracking, which shouldn't have this identical streamline situation unless something is wrong with the way the bundle was generated (e.g. duplicated samples), in which case the user needs to address that to avoid getting misleading results)

Does this sound reasonable?

Reply:
yep, good for me. Thank you for your explanation